### PR TITLE
[v5] plumbing: format/index, Improve v4 entry name validation

### PR DESCRIPTION
**Pull request by pjbgf**
Updates Git index (`plumbing/format/index`) encoding/decoding logic to better match on-disk format edge cases (long entry names and V4 name prefix-compression), and adds regression tests around these behaviors.

Backport of #1934.

### CODE REVIEW

File: plumbing/format/index/encoder.go [hunk 1]
```diff
@@ -160,26 +158,39 @@ func (e *Encoder) encodeEntryName(entry *Entry) error {
...
+// commonPrefixLen returns the length of the longest common byte prefix
+// between a and b.
+func commonPrefixLen(a, b string) int {
+	n := min(len(b), len(a))
+	for i := range n {
+		if a[i] != b[i] {
+			return i
+		}
+	}
+	return n
```
[the remark this answers is not in the document]
Comment (pjbgf) [reply plumbing/format/index/encoder.go:193]: This is incorrect - the code does compile.
